src/nvim/eval.c

```diff
@@ -11686,14 +11691,70 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
         return;
       }
     }
+    dictitem_T *item;
+    // or new_env
+    item = tv_dict_find(job_opts, S_LEN("env"));
+    if (item) {
+      size_t env_size = 0; // len ?
+      size_t i = 0;
+      // size_t orig_env_size = 0;
+      if (item->di_tv.v_type != VAR_DICT) {
+        EMSG2(_(e_invarg2), "Expected a dictionary for 'env'");
+        return;
+      }
+
+      // number of items in dict
+      env_size += (size_t)tv_dict_len(item->di_tv.vval.v_dict);
+      ILOG("%d passed env variables", env_size);
+
+      if (!reset_env) {
+        char **environ = os_getfullenv();
+        size_t current_env_size = 0; // = strlen(environ); // strlen excludes null byte
+        for (env = environ; *env; env++) {
+          current_env_size++;
+        }
+        ILOG("current env_len=%d", current_env_size);
+        env_size += current_env_size;
+        env = xmalloc( (env_size) * sizeof(char));
+        for (;i < current_env_size; i++) {
+            env[i] = xstrdup(environ[i]);
+            ILOG("from env: %s (id=%d)", env[i], i);
+        }
+      } else {
+        // + 1 for last null entry
+        env = xmalloc( (env_size + 1) * sizeof(char));
+      }
+      assert(env);  // env should be allocated here
+      ILOG("total env_len=%d", env_size);
+
+      // TODO add the following in tv_to_argv  ?
+      TV_DICT_ITER(item->di_tv.vval.v_dict, var, {
+          // tv_get_string => single buffer vs tv_get_string_buf
+          const char *str = tv_get_string(&var->di_tv);
+          if (str) {
+            size_t len = STRLEN(var->di_key) + strlen(str) + strlen("=") + 1;
+            ILOG("value #%d: %s", i, str);
+            env[i] = xmallocz(len * sizeof(char));
```

Comment:
Also you allocated one more bytes then needed: +1 in `len` calculation is for NUL I guess, but `xmallocz` already allocates one more bytes for NUL. Also given that you use `snprintf` `xmallocz` is not needed, `snprintf` will add bytes itself, so just replace code with `env[i] = xmalloc(len);`.